### PR TITLE
helm: Enable Hubble mTLS in experimental-install.yaml via K8s CronJob

### LINKED ISSUE #13233: Hubble: provide alternative to Helm for automatic TLS certs generation
PR #12906 and #12900 introduced mTLS support for Hubble server and Hubble Relay. This support comes with automatic TLS certificates creation via Helm and is enabled by default when deploying Cilium with Hubble and Hubble Relay.

There are two caveats with using Helm to generate TLS certificates:

- The Helm templates make use of [sprig's `genCA` and `genSignedCert` functions](https://masterminds.github.io/sprig/crypto.html) which offer very limited configuration options. For instance, one cannot use ECDSA keys or specify the size of an RSA key.
- It is not possible to enable mTLS automatically when not using Helm. Typically, the `experimental-install.yml` installation file disables mTLS altogether. Users who do not use Helm but want to enable mTLS have to manually provide certificates, which is not ideal and creates some friction (and worse: it encourages them not to enable mTLS between Hubble and Hubble Relay).

For the reasons mentioned above (and especially the 2nd one), we should explore an alternative way of automatically generating TLS certificates.

### PR DESCRIPTION
**Pull request by gandro**
This adds a new method to automatically generate mTLS certificates for
Hubble at run-time without relying on Helm. This for example is useful
when a pregenerated YAML manifest (e.g.  `experimentall-install.yaml`)
which with the Helm based would re-use the same set of certificates for
all installations.

This new approach generates the certificates at run-time instead. When
`hubble.tls.auto.method` is set to `cronJob`, the generated YAML will
create a Kubernetes [Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) to generate the certificates at installation
time. The job is running in host namespace and will store the
certificates as Kubernetes secrets. Once the secerts are creatd,
Kubernetes will automatically mount the generated certificates into the
already running Cilium pod, allowing it to pick them up and start
serving the Hubble API with mTLS (c.f. #13249). See commit
message of f4e2cae for more details.

In addition to the one-shot job to generate the initial set of
certificates, an additional recurring Kubernetes CronJob is deployed to
regenerate the certificate in a regular interval (regardless of the
expiration date). This cronjob can be optionally disabled in Helm by
unsetting `hubble.tls.auto.cronJob.schedule`.
Existing connections are not affected by the certificate or CA renewal.

The source code for the cron job docker image can be found at
https://github.com/cilium/certgen.

```release-note
Add an alternative method to generate the Hubble mTLS certificates based on Kubernetes Jobs.
```

Fixes: #13233

### CODE REVIEW

**Comment by gandro** [comment]
Rebased. The correct generation of `experimental-install.yaml` is currently blocked on #13523

**Comment by gandro** [comment]
test-me-please

File: install/kubernetes/cilium/templates/hubble-generate-certs-clusterrole.yaml
Comment (aanm) [inline install/kubernetes/cilium/templates/hubble-generate-certs-clusterrole.yaml]:
```
{{- if and .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "cronJob") .Values.serviceAccounts.certgen.create }}
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: hubble-generate-certs
rules:
  - apiGroups:
      - ""
    resources:
      - secrets
    resourceNames:
      - hubble-ca-secret
    verbs:
      - create
      - update
  - apiGroups:
      - ""
    resources:
      - configmaps
    resourceNames:
      - hubble-ca-cert
    verbs:
      - create
      - update
{{- end }}
```

File: install/kubernetes/cilium/values.yaml
Comment (aanm) [inline install/kubernetes/cilium/values.yaml]: `s/Always/IfNotPresent` (we have a pin tag)

File: install/kubernetes/cilium/values.yaml
Comment (aanm) [inline install/kubernetes/cilium/values.yaml]: Since this is a CronJob I would suggest to set the sha256 part of the tag to avoid accidentally pushes to the `v0.1.0` and people assuming the `v0.1.0` is still safe.

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by gandro** [comment]
Pushed a new version which actually addresses @aanm's [feedback](https://github.com/cilium/cilium/pull/13449#discussion_r505545797) to set the `resourceNames` for the `update` permission in the `ClusterRole`.

`create` remains unrestricted, as it is not supported by K8s ([source](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)):
> Note: You cannot restrict create or delete collection requests by resourceName. For create, this limitation is because the object name is not known at authorization time.

**Review by aanm — APPROVED** [review]

**Review by joestringer — APPROVED** [review]
I think this is pulling in @cilium/docs-structure  due to `install/kubernetes` changes which could be better reviewed by Hubble folks (particularly those involved in the TLS work). 

`Makefile` changes LGTM. I also recognize that no changes are being made to `quick-install.yaml` and the GH actions passed, so this seems unlikely to impact folks unless they tweak the TLS settings. But anyway I'll give the :+1: for docs team and defer to hubble to review & merge.

**Review by sayboras — APPROVED** [review]
LGTM :100: 

As this https://github.com/cilium/cilium/pull/13584 is just merged, do you mind to rebase with master, and run `make -C install/kubernetes docs` to re-generate README.md ? Thanks :+1:

**Comment by gandro** [comment]
Rebased to pick up the README update

**Comment by gandro** [comment]
test-me-please

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml
Comment (rolinh) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:497]: nit: it's easier to read/parse when `optional` is just below `name` rather than the `items` list.

File: install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl
Comment (rolinh) [inline install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl:61]: nit: For consistency with other TLS secrets (ie: `hubble-server-certs`, `hubble-relay-client-certs`, `hubble-relay-server-certs`), I think the Hubble CA secret should be named `hubble-ca-certs` rather than `hubble-ca-secret`.

File: install/kubernetes/cilium/templates/hubble-generate-certs-ca-secret.yaml
Comment (rolinh) [inline install/kubernetes/cilium/templates/hubble-generate-certs-ca-secret.yaml:6]: ditto

**Review by rolinh — APPROVED** [review]
This is awesome! I reviewed and tested different scenario and everything seems to work as expected. :rocket: LGTM
The nits below need not to be addressed.

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml
Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:497]: Many of the Cilium yaml definitions seem to use the alphabetical order. That's why optional is where it is.

File: install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl
Comment (gandro) [reply install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl:61]: I intentionally did not choose `hubble-ca-certs` to avoid confusion with 
https://github.com/cilium/cilium/blob/e9cb43c03179af6f7cd5d0abb9974daeb5a21e5a/install/kubernetes/cilium/templates/hubble-ca-configmap.yaml#L5 :(

Comment (rolinh) [reply install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl:61]: I see. But `hubble-ca-cert` is a configmap, not a secret. Thus, I think it makes even more sense to name it `hubble-ca-certs` because the cert value in the configmap and the secret should actually be the same.

Comment (gandro) [reply install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl:61]: I think I disagree. I personally find it more confusing and potentially error prone if they would share the same name. 

But in the end, personal preference is not the argument to make here, consistency with the rest of the existing code base should take priority over personal preferences.

I have not yet checked if there is prior art for this pattern in the rest of of Cilium, but I'm marking this as resolved for this PR: If we decide to use the same name, then we will have to rename the above config map as well, and that is a change unrelated to this PR. So if we decide to rename, then let's do that in a follow-up.

**Comment by gandro** [comment]
> This is awesome! I reviewed and tested different scenario and everything seems to work as expected. rocket LGTM
> The nits below need not to be addressed.

Thanks a lot! As mentioned inline, I will not address them in this PR. Let's discuss the naming in a follow up instead, as it likely requires some unrelated changes.

**Comment by aanm** [comment]
retest-4.9

**Comment by aanm** [comment]
retest-4.9

**Pull request merged**
